suite/tests/common/decode.c

```diff
@@ -145,6 +145,90 @@ actual_call_target(void)
     print("Made it to actual_call_target\n");
 }
 
+#    ifndef X64
+/****************************************************************************
+ * Macros and routine to construct the encoded instruction buffer for the
+ * modrm test.
+ */
+#        define NOP_ENC 0x90
+#        define PROLOG_SIZE 2   /* save esp */
+#        define TEST_SEQ_SIZE 6 /* actual test seq. */
+#        define EPILOG_SIZE 3   /* restore esp, return */
+#        define PROLOG_START 0
+#        define TEST_SEQ_START (PROLOG_START + PROLOG_SIZE)
+#        define EPILOG_START (TEST_SEQ_START + TEST_SEQ_SIZE)
+#        define EACH_SEQ_SIZE (EPILOG_START + EPILOG_SIZE)
+#        define TOTAL_BUF_SIZE (EACH_SEQ_SIZE * 256 + 1)
+
+static void
+construct_modrm_test_buf(char *buf)
+{
+    /* Add an encoded instr for each of the 256 variants of the modr/m
+     * byte. Each of these instructions is part of a sequence of instrs:
+     * prolog -> modrm instr -> epilog -> ret
+     * nop may be added so that each of these parts have consistent size
+     * for all 256 variants.
+     */
+    for (int j = 0; j < 256; j++) {
+        int mod = ((j >> 6) & 0x3); /* top 2 bits */
+        int reg = ((j >> 3) & 0x7); /* middle 3 bits */
+        int rm = (j & 0x7);         /* bottom 3 bits */
+
+        /* Set prolog and epilog. */
+        if (IF_X64_ELSE(false, reg == 4 /* esp */)) {
+            /* We spill esp to a reg and restore it after the test sequence.
+             * Without this, esp will get clobbered and ret may segfault.
+             * Not yet ported to X86-64 as this test is disabled there.
+             */
+            buf[j * EACH_SEQ_SIZE + PROLOG_START + 0] = 0x89; /* mov */
+            buf[j * EACH_SEQ_SIZE + EPILOG_START + 0] = 0x89; /* mov */
+            if (mod == 3 && rm == 0) {
+                /* As eax is a source reg, we use ebx as the save slot instead. */
+                buf[j * EACH_SEQ_SIZE + PROLOG_START + 1] = 0xe3; /* esp -> ebx */
+                buf[j * EACH_SEQ_SIZE + EPILOG_START + 1] = 0xdc; /* ebx -> esp */
+            } else {
+                /* Use eax to save esp in prolog and restore in epilog. */
+                buf[j * EACH_SEQ_SIZE + PROLOG_START + 1] = 0xe0; /* esp -> eax */
+                buf[j * EACH_SEQ_SIZE + EPILOG_START + 1] = 0xc4; /* eax -> esp */
+            }
+        } else {
+            /* Prolog and epilog are not needed for non-xsp dest, as we can
+             * handle other regs being clobbered.
+             */
+            buf[j * EACH_SEQ_SIZE + PROLOG_START + 0] = NOP_ENC;
```

Comment:
So these nops are only here for each entry to have the exact same size and simplify sizing the buffer up front?  Maybe worth a comment.

Reply:
It's already there on L169, at the beginning of `construct_modrm_test_buf`.

Reply:
Added the part about it helping to simplify buf size up front.